mobile_config/src/key_cache.rs

```diff
@@ -0,0 +1,195 @@
+            Ok(())
+        } else {
+            Err(anyhow!("unauthorized request"))
+        }
+    }
+
+    pub fn verify_signature_with_type<R>(
+        &self,
+        key_type: KeyType,
+        signer: &PublicKey,
+        request: &R,
+    ) -> anyhow::Result<()>
+    where
+        R: MsgVerify,
+    {
+        if self.cache_receiver.borrow().get_key_value(signer) == Some((signer, &key_type))
+            && request.verify(signer).is_ok()
+        {
+            tracing::debug!(pubkey = signer.to_string(), "request authorized");
+            Ok(())
+        } else {
+            Err(anyhow!("unauthorized request"))
+        }
+    }
+
+    pub fn get_keys(&self) -> Vec<(PublicKey, KeyType)> {
+        self.cache_receiver
+            .borrow()
+            .iter()
+            .map(|(k, t)| (k.clone(), *t))
+            .collect()
+    }
+
+    pub fn get_keys_by_type(&self, key_type: KeyType) -> Vec<PublicKey> {
+        self.cache_receiver
+            .borrow()
+            .iter()
+            .filter_map(|(k, t)| {
+                if t == &key_type {
+                    Some(k.clone())
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, Serialize, sqlx::Type)]
+#[sqlx(type_name = "key_type", rename_all = "snake_case")]
+pub enum KeyType {
+    Administrator,
+    Oracle,
+    PacketRouter,
+}
+
+impl KeyType {
+    pub fn from_i32(v: i32) -> anyhow::Result<Self> {
+        ProtoKeyType::from_i32(v)
+            .map(|kt| kt.into())
+            .ok_or(anyhow!("unsupported key type {}", v))
+    }
+}
+
+impl From<KeyType> for ProtoKeyType {
+    fn from(key_type: KeyType) -> Self {
+        ProtoKeyType::from(&key_type)
+    }
+}
+
+impl From<&KeyType> for ProtoKeyType {
+    fn from(skt: &KeyType) -> Self {
+        match skt {
+            KeyType::Administrator => ProtoKeyType::Administrator,
+            KeyType::Oracle => ProtoKeyType::Oracle,
+            KeyType::PacketRouter => ProtoKeyType::PacketRouter,
+        }
+    }
+}
+
+impl From<ProtoKeyType> for KeyType {
+    fn from(kt: ProtoKeyType) -> Self {
+        match kt {
+            ProtoKeyType::Administrator => KeyType::Administrator,
+            ProtoKeyType::Oracle => KeyType::Oracle,
+            ProtoKeyType::PacketRouter => KeyType::PacketRouter,
+        }
+    }
+}
+
+impl std::fmt::Display for KeyType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let s = match self {
+            Self::Administrator => "administrator",
+            Self::Oracle => "oracle",
+            Self::PacketRouter => "packet_router",
+        };
+        f.write_str(s)
+    }
+}
+
+pub(crate) mod db {
+    use super::{KeyType, PublicKey, PublicKeyBinary};
+    use sqlx::Row;
+
+    pub(crate) async fn fetch_stored_keys(
+        db: impl sqlx::PgExecutor<'_>,
+    ) -> anyhow::Result<Vec<(PublicKey, KeyType)>> {
```

Comment:
You can just have this return a `CachedKey` instead of into_itering the result and re-collecting

Reply:
good call; i probably _thought_ i wanted a generic result here at one point but i'm never using this function anywhere else so that's a good recommendation